File: Python/Valid Parentheses.py
"""
Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.
"""

class Solution:
    # @return a boolean
    def isValid(self, s):
        stack = []
        M = len(s)
        while M > 0:
            if len(stack) == 0:
                stack.append(s[len(s)-M])
                M-=1
            else:
                c = stack.pop()
                tmp = s[len(s) - M]
                if c is '(' and tmp is ')' or c is '[' and tmp is ']' or c is '{' and tmp is '}':
                    M-=1
                else:
                    stack.append(c)
                    stack.append(s[len(s)-M])
                    M -=1
        if len(stack) == 0:
            return True
        else:
            return False

File: Java/Valid Parentheses.java
/*
Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.
*/

public class Solution {
    private static final Map<Character, Character> map =
      new HashMap<Character, Character>() {{
            put('(', ')');
            put('{', '}');
            put('[', ']');
    }};  
    
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(char c:s.toCharArray()){
            if(map.containsKey(c))
                stack.push(c);
            else if(stack.isEmpty() || map.get(stack.pop()) != c)
                return false;
        }
        return stack.isEmpty();
    }
}